quic/packets.nim

```diff
@@ -1,83 +1,178 @@
 import math
 import strutils
-import strformat
 import bits
 
 type
-  HeaderForm* = enum
-    headerShort
-    headerLong
+  ConnectionId* = distinct seq[byte]
+  PacketForm* = enum
+    formShort
+    formLong
   PacketKind* = enum
     packetInitial
     packet0RTT
     packetHandshake
     packetRetry
     packetVersionNegotiation
-  PacketHeader* = object
-    bytes: seq[byte]
+  HeaderInitial* = object
+    version*: uint32
+  Header0RTT* = object
+    version*: uint32
+  HeaderHandshake* = object
+    version*: uint32
+  HeaderRetry* = object
+    version*: uint32
+  HeaderVersionNegotiation* = object
+    supportedVersion*: uint32
+  Packet* = object
+    case form*: PacketForm
+    of formShort:
+      discard
+    of formLong:
+      case kind*: PacketKind
+      of packetInitial: initial*: HeaderInitial
+      of packet0RTT: rtt*: Header0RTT
+      of packetHandshake: handshake*: HeaderHandshake
+      of packetRetry: retry*: HeaderRetry
+      of packetVersionNegotiation: negotiation*: HeaderVersionNegotiation
+      destination*: ConnectionId
+      source*: ConnectionId
   PacketNumber* = range[0'u64..2'u64^62-1]
-  ConnectionId* = distinct seq[byte]
 
-proc newPacketHeader*(bytes: seq[byte]): PacketHeader =
-  assert bytes[0].bits[1] == 1
-  PacketHeader(bytes: bytes)
+proc readForm(datagram: seq[byte]): PacketForm =
+  PacketForm(datagram[0].bits[0])
+
+proc writeForm(datagram: var seq[byte], header: Packet) =
+  datagram[0].bits[0] = Bit(header.form)
 
-proc bytes*(header: PacketHeader): seq[byte] =
-  header.bytes
+proc readFixedBit(datagram: seq[byte]) =
+  assert datagram[0].bits[1] == 1
```

Comment:
we use `doAssert` so that we get the checks in all compile modes

Reply:
Resolved in 1d675a7